### DISCUSSION TITLE
Letters in decreasing order of frequency

### DISCUSSION THREAD

**Discussion by MaziarAm** [community/community#21410] [Other Feature Feedback, Questions, & Ideas]
<p>Hey everone, so I had this exercise :</p>
<p>“Write a program that reads a file and prints the letters<br>
in decreasing order of frequency. Your program should convert all the<br>
input to lower case and only count the letters a-z. Your program should<br>
not count spaces, digits, punctuation, or anything other than the letters<br>
a-z.”</p>
<p>And I managed to solve it, and here’s a part of my code :</p>
<pre><code>fname = input("Enter a file name: ")
if len(fname) &lt; 1 : fname = "romeo.txt"
try :
    fhand = open(fname)
except :
    print("Enter a proper file name",fname )
    quit()

#read the file into a big string
words = fhand.read()

#lower the string
words = words.lower()

letters = list()
for i in words :
    if i.isalpha() == True :
        letters.append(i)
</code></pre>
<p>And it works just fine. However, I was looking up for other solutions online and I found one in the answer section of the book(Python for everybody), it’s pretty much like mine but it has an extra line :</p>
<pre><code>words = fhand.read()

#lower the string, don't worry about digits, punctuation etc.
words = words.lower()

#change the string into single letters
twords = tuple(words)

letters = list()
for i in twords :
#we only choose the alphabet letters
    if i.isalpha() == True :
        letters.append(i)
</code></pre>
<p>I was wondering why the following line is added ? Because the code just works fine without it </p>
<pre><code>twords = tuple(words)
</code></pre>

**Comment by fire-eggs** [comment]
<p>Without being able to absolutely confirm this, I believe this is a historical oddity. The behavior of strings changed between Python 2 and Python 3; the tuple() line is likely necessary for Python 2 support with Unicode?</p>

**Comment by Wabri** [comment, marked as answer]
<p>I dont think there is a real reason. I think the only one reason that i have in mind is maybe to split the tasks of words and twords. The output of letters with twords and words are the same (I didnt tried it but I’m sure of that). To test the output of words and twords you can do something like this:</p>
<pre><code>fname = input("Enter a file name: ")
if len(fname) &lt; 1 : fname = "romeo.txt"
try :
    fhand = open(fname)
except :
    print("Enter a proper file name",fname )
    quit()

#read the file into a big string
words = fhand.read()

#lower the string
words = words.lower()

letters = list()
for i in words :
    if i.isalpha() == True :
        letters.append(i)

twords = tuple(words)

letters2 = list()
for i in twords :
    if i.isalpha() == True :
        letters2.append(i)

print(letters==letters2)
</code></pre>
<p>and if the output is True than the output are the same (I’m sure it will).</p>
<p>Hope to solve your doubts</p>
<p>-Gabriele-</p>

**Comment by MaziarAm** [comment]
<p>Thanks!</p>
<p>I’m not sure but maybe it’s what you’re saying because in some of the book slides,there are references to Python 2.</p>

**Comment by MaziarAm** [comment]
<p>Yeah I think it’s used maybe to explain a little more about what’s going on. And yes the output is exactly the same</p>
<p>Thanks !</p>

**Comment by Wabri** [comment]
<p>Dont forget to set it as solution :slightly_smiling_face:</p>

**Comment by pythonlearner18** [comment]
<p>Where did you count the frequency of alphabets?</p>

**Comment by vmatom** [comment]
<p>Well here is no count. But it’s easy to add.</p>
<pre><code class="lang-auto">counts = dict() #create empty dictionary
disallowed_characters="!,.?:;'—...\n1234567890()-/+&gt;&lt;#@$%^&amp;*~№| " #add line of punctuation marks that we need to expel
try:
    op = open(input("Enter a file name:")).read().lower()
except FileNotFoundError:
    print('File cannot be opened:', fname)
    quit()
for character in disallowed_characters:
    op= op.replace(character,'')
#print(op)
#type(op)
letters=list()
#print(letters)
for i in op:
    counts[i]=counts.get(i,0)+1
#print(counts)
for k,v in counts.items():
    letters.append((v,k))
letters=sorted(letters,reverse=True)
for v,k in letters:
    print (k,v)
</code></pre>

**Comment by ghost** [comment]
Oh, everything turns out to be so simple, but I solved it using the methods of the py4e books + a little help on github, because I could not fully understand the assignment


```python

import string

fileName = input('Enter a file name: ')
file = open(fileName)

count = 0
d = dict()
lst = list()

for line in file:
    line = line.translate(line.maketrans('', '', string.digits))
    line = line.translate(line.maketrans('', '', string.punctuation))
    line = line.lower()

    words = line.split()
    for word in words:
        for letter in word:
            count += 1
            d[letter] = d.get(letter, 0) + 1

for k, v in list(d.items()):
    lst.append((v / count, k))

lst.sort(reverse=True)

for k, v in lst:
    print(f'{v} {k:.3f}')
```
